src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -51,7 +51,7 @@ export function absThreshold(diff: number): FloatMatch {
  */
 export function ulpThreshold(ulp: number): FloatMatch {
   return (got, expected) => {
-    if (got === expected) {
+    if (got === expected || (Number.isNaN(got) && Number.isNaN(expected))) {
```

Comment:
This is because comparing NaN to NaN is always false. I am not sure if this is behaviour we want all the time for tests, or if there should be a NaN-y and non-NaN-y version of ulpThreshold.